radix-engine/src/system/transaction/multithread_intent_processor.rs

```diff
@@ -0,0 +1,189 @@
+            let auth_zone = AuthModule::create_auth_zone(
+                &mut system_service,
+                None,
+                simulate_every_proof_under_resources,
+                initial_non_fungible_id_proofs,
+            )?;
+
+            api.kernel_set_call_frame_data(Actor::Function(FunctionActor {
+                blueprint_id: BlueprintId::new(
+                    &TRANSACTION_PROCESSOR_PACKAGE,
+                    TRANSACTION_PROCESSOR_BLUEPRINT,
+                ),
+                ident: TRANSACTION_PROCESSOR_RUN_IDENT.to_string(),
+                auth_zone,
+            }))?;
+
+            let mut system_service = SystemService::new(api);
+            let txn_processor = IntentProcessor::<InstructionV2>::init(
+                intent.encoded_instructions.clone(),
+                global_address_reservations.clone(),
+                intent.blobs.clone(),
+                MAX_TOTAL_BLOB_SIZE_PER_INVOCATION,
+                &mut system_service,
+            )?;
+
+            txn_processors.push((
+                txn_processor,
+                intent
+                    .children_subintent_indices
+                    .iter()
+                    .map(|index| {
+                        let thread_index = index.0 + 1;
+                        thread_index
+                    })
+                    .collect::<Vec<_>>(),
+            ));
+        }
+        Ok(Self {
+            threads: txn_processors,
+        })
+    }
+
+    pub fn execute<Y: SystemBasedKernelApi>(&mut self, api: &mut Y) -> Result<(), RuntimeError> {
+        let mut cur_thread = 0;
+        let mut parent_stack = vec![];
+        let mut passed_value = None;
+
+        loop {
+            api.kernel_switch_stack(cur_thread)?;
+            let (txn_thread, children_mapping) = self.threads.get_mut(cur_thread).unwrap();
+
+            let mut system_service = SystemService::new(api);
+            let switch_to = match txn_thread.resume(passed_value.take(), &mut system_service)? {
+                ResumeResult::YieldToChild(child, value) => {
+                    let child = *children_mapping.get(child).unwrap();
+                    parent_stack.push(cur_thread);
+                    Some((child, value, false))
+                }
+                ResumeResult::YieldToParent(value) => {
+                    let parent = parent_stack.pop().unwrap();
+                    Some((parent, value, false))
+                }
+                ResumeResult::ChildIntentDone(value) => {
+                    let parent = parent_stack.pop().unwrap();
+                    Some((parent, value, true))
+                }
+                ResumeResult::RootIntentDone => None,
+            };
+
+            if let Some((next_thread, value, intent_done)) = switch_to {
+                // Checked passed values
+                Self::check_yielded_value(&value, api)?;
+                api.kernel_send_to_stack(next_thread, value.clone())?;
```

Comment:
Why do we both need to `send_to_stack` and send value to `txn_thread.resume(..)`?

I guess `send_to_stack` is about handling moving the owned objects, but we need to tell the other thread what it receives?

Can we avoid the clone if `kernel_send_to_stack` takes a reference to the value? I assume it just needs to extract references from it?

Reply:
Yes exactly. It's similar to our logic of calling a method with arguments, we need to update kernel call frame as well as let higher level understand what the args are. The difference here is that "Threads" is a system construct rather than kernel construct (If we implemented kernel threads the args can come from the kernel via callback).

Yup, good point, will update to use reference.